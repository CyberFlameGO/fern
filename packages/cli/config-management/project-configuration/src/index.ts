export {
    DEFINITION_DIRECTORY,
    FERN_DIRECTORY,
    GENERATORS_CONFIGURATION_FILENAME,
    PROJECT_CONFIG_FILENAME,
    ROOT_API_FILENAME,
} from "./constants";
export { getPathToFernDirectory } from "./getPathToFernDirectory";
export { loadProjectConfig, type ProjectConfig as ProjectConfig } from "./loadProjectConfig";
export { ProjectConfigSchema } from "./schemas/ProjectConfigSchema";
