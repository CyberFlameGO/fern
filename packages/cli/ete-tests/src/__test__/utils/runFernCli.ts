import execa, { Options } from "execa";
import path from "path";

export async function runFernCli(args: string[], options?: Options): Promise<execa.ExecaChildProcess> {
    const cmd = execa("node", [path.join(__dirname, "../../../../cli/webpack/dist/bundle.js"), ...args], options);
    cmd.stdout?.pipe(process.stdout);
    cmd.stderr?.pipe(process.stderr);
    return cmd;
}
