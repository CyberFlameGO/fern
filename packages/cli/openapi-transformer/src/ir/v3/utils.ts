import { FernOpenapiIr } from "@fern-fern/openapi-ir-sdk";
import { OpenAPIV3 } from "openapi-types";

export const APPLICATION_JSON_CONTENT = "application/json";
export const SCHEMA_REFERENCE_PREFIX = "#/components/schemas/";
export const RESPONSE_REFERENCE_PREFIX = "#/components/responses/";
export const REQUEST_REFERENCE_PREFIX = "#/components/requests/";

export function isReferenceObject(
    parameter:
        | OpenAPIV3.ReferenceObject
        | OpenAPIV3.ParameterObject
        | OpenAPIV3.SchemaObject
        | OpenAPIV3.RequestBodyObject
        | OpenAPIV3.SecuritySchemeObject
): parameter is OpenAPIV3.ReferenceObject {
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    return (parameter as OpenAPIV3.ReferenceObject).$ref != null;
}

export function getSchemaIdFromReference(reference: OpenAPIV3.ReferenceObject): FernOpenapiIr.SchemaId | undefined {
    if (reference.$ref.startsWith(SCHEMA_REFERENCE_PREFIX)) {
        return reference.$ref.substring(SCHEMA_REFERENCE_PREFIX.length);
    }
    return undefined;
}
