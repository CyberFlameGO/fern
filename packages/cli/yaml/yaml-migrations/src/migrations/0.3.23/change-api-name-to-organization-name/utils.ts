import { AbsoluteFilePath, join } from "@fern-api/fs-utils";
import { TaskContext } from "@fern-api/task-context";
import { findUp } from "find-up";
import { readFile } from "fs/promises";
import glob from "glob-promise";

const FERN_DIRECTORY = "fern";
const FERN_ROOT_CONFIG_FILENAME = "fern.config.json";

export async function getOrganizationName(context: TaskContext): Promise<string> {
    const fernDirectory = await getFernDirectory();
    if (fernDirectory == null) {
        context.failAndThrow(`Directory "${FERN_DIRECTORY}" not found.`);
    }
    const fernConfigJsonBuffer = await readFile(join(fernDirectory, FERN_ROOT_CONFIG_FILENAME));
    const fernConfigJson = JSON.parse(fernConfigJsonBuffer.toString());
    const organizationName = fernConfigJson.organization;
    if (organizationName == null) {
        context.failAndThrow(`${FERN_ROOT_CONFIG_FILENAME} is missing organization`);
    }
    return organizationName;
}

export async function getAllApiYmlFiles(context: TaskContext): Promise<AbsoluteFilePath[]> {
    const fernDirectory = await getFernDirectory();
    if (fernDirectory == null) {
        context.failAndThrow(`Directory "${FERN_DIRECTORY}" not found.`);
    }
    const filepaths = await glob("*/definition/api.yml", {
        cwd: fernDirectory,
        absolute: true,
    });
    return filepaths.map(AbsoluteFilePath.of);
}

async function getFernDirectory(): Promise<AbsoluteFilePath | undefined> {
    const fernDirectoryStr = await findUp(FERN_DIRECTORY, { type: "directory" });
    if (fernDirectoryStr == null) {
        return undefined;
    }
    return AbsoluteFilePath.of(fernDirectoryStr);
}
