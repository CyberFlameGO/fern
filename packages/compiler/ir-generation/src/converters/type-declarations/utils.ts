import { RawSchemas } from "@fern-api/syntax-analysis";

export function isRawAliasDefinition(
    rawTypeDeclaration: RawSchemas.TypeDeclarationSchema
): rawTypeDeclaration is RawSchemas.AliasSchema {
    return (rawTypeDeclaration as RawSchemas.AliasSchema)?.alias != null;
}

export function isRawObjectDefinition(
    rawTypeDeclaration: RawSchemas.TypeDeclarationSchema
): rawTypeDeclaration is RawSchemas.ObjectSchema {
    return (rawTypeDeclaration as RawSchemas.ObjectSchema)?.properties != null;
}

export function isRawUnionDefinition(
    rawTypeDeclaration: RawSchemas.TypeDeclarationSchema
): rawTypeDeclaration is RawSchemas.UnionSchema {
    return (rawTypeDeclaration as RawSchemas.UnionSchema)?.union != null;
}

export function isRawEnumDefinition(
    rawTypeDeclaration: RawSchemas.TypeDeclarationSchema
): rawTypeDeclaration is RawSchemas.EnumSchema {
    return (rawTypeDeclaration as RawSchemas.EnumSchema)?.enum != null;
}
