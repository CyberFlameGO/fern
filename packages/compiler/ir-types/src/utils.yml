# These types aren't actually present in the IR, but may be helpful
# for generators

imports:
  types: ./types.yml
  errors: ./errors.yml

types:
  ModelReference:
    union:
      type: types.DeclaredTypeName
      error: errors.ErrorName
