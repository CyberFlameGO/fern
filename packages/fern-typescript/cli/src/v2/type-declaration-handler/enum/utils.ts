import { EnumValue } from "@fern-fern/ir-model";
import snakeCase from "lodash-es/snakeCase";
import upperFirst from "lodash-es/upperFirst";

export function getKeyForEnum({ name }: EnumValue): string {
    return snakeCase(name).split("_").map(upperFirst).join("");
}

export function getValueForEnum({ value }: EnumValue): string {
    return snakeCase(value).toUpperCase();
}
