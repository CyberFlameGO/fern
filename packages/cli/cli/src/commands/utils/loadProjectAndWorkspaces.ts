import { loadProjectConfig, ProjectConfig } from "@fern-api/project-configuration";
import { getUniqueWorkspaces } from "./getUniqueWorkspaces";

export async function loadProjectAndWorkspaces({
    commandLineWorkspaces,
}: {
    commandLineWorkspaces: readonly string[];
}): Promise<{
    projectConfig: ProjectConfig;
    workspacePaths: string[];
}> {
    const projectConfig = await loadProjectConfig();

    const workspacePaths = await getUniqueWorkspaces({
        commandLineWorkspaces,
        projectConfig,
    });

    return { projectConfig, workspacePaths };
}
