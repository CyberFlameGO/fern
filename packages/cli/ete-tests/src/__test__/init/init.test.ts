import { PROJECT_CONFIG_FILENAME } from "@fern-api/project-configuration";
import { WORKSPACE_CONFIGURATION_FILENAME } from "@fern-api/workspace-configuration";
import { mkdir, readFile, rm } from "fs/promises";
import path from "path";
import { runFernCli } from "../utils/runFernCli";

const GENERATED_DIR = path.join(__dirname, "generated");
const GENERATED_API_DIR = path.join(GENERATED_DIR, "api");

it("fern init", async () => {
    await rm(GENERATED_DIR, { force: true, recursive: true });
    await mkdir(GENERATED_DIR);

    await runFernCli(["init", "--organization", "fern"], {
        cwd: GENERATED_DIR,
    });

    expect(await getFileContentsAsString(path.join(GENERATED_DIR, PROJECT_CONFIG_FILENAME))).toMatchSnapshot();
    expect(
        await getFileContentsAsString(path.join(GENERATED_API_DIR, WORKSPACE_CONFIGURATION_FILENAME))
    ).toMatchSnapshot();
    expect(await getFileContentsAsString(path.join(GENERATED_API_DIR, "src", "api.yml"))).toMatchSnapshot();
}, 60_000);

async function getFileContentsAsString(filepath: string) {
    const fileContents = await readFile(filepath);
    return fileContents.toString();
}
