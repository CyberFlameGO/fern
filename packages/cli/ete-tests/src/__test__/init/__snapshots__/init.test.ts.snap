// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fern init 1`] = `
"{
    \\"workspaces\\": [
        \\"**\\"
    ],
    \\"organization\\": \\"fern\\"
}"
`;

exports[`fern init 2`] = `
"name: api
definition: src
generators: []
"
`;

exports[`fern init 3`] = `
"# yaml-language-server: $schema=https://raw.githubusercontent.com/fern-api/fern/main/fern.schema.json

# This is a sample IMDb API to get you more familiar with defining APIs in Fern.

types:
  MovieId: string
  Movie:
    properties:
      id: MovieId
      title: string
      rating: double

  CreateMovieRequest: 
    properties:
      title: string
      rating: double 

services:
  http:
    MoviesService:
      auth: none
      base-path: /movies
      endpoints:

        # Here's an HTTP endpoint. Fern uses sane defaults for endpoint path and HTTP method.
        createMovie:
          request: CreateMovieRequest
          response: MovieId

        getMovie:
          request: MovieId
          response: Movie
          errors:
            - NotFoundError

errors:
  NotFoundError:
    http:
      statusCode: 404"
`;
